pkg/database/database.go

```diff
@@ -14,238 +15,169 @@ import (
+func (d *Database) ensureModels(dctx *Context) error {
+	namespace := *dctx.namespace
+	tenant, err := tenantsmodel.Ensure(dctx, dctx.transaction, &db.Tenant{Source: *dctx.organization})
 	if err != nil {
 		return fmt.Errorf("cannot ensure organization for namespace %s: %w", namespace, err)
 	}
-	ctx.tenant = tenant
+	dctx.tenant = tenant
 
-	category, err := categoriesmodel.Ensure(ctx, ctx.transaction, &db.Category{Source: provider + ":" + namespace})
+	category, err := categoriesmodel.Ensure(dctx, dctx.transaction, &db.Category{Source: provider + ":" + namespace})
 	if err != nil {
 		return fmt.Errorf("cannot ensure category for namespace %s: %w", namespace, err)
 	}
-	ctx.category = category
+	dctx.category = category
 
-	dateTime := datetimesmodel.New(ctx.billingDate)
-	dateTime, err = datetimesmodel.Ensure(ctx, ctx.transaction, dateTime)
+	dateTime := datetimesmodel.New(d.BillingDate)
+	dateTime, err = datetimesmodel.Ensure(dctx, dctx.transaction, dateTime)
 	if err != nil {
 		return fmt.Errorf("cannot ensure date time for namespace %s: %w", namespace, err)
 	}
-	ctx.dateTime = dateTime
+	dctx.dateTime = dateTime
 	return nil
 }
 
-func (d *Database) getBestMatch(ctx *transactionContext) error {
-	namespace := *ctx.namespace
-	productMatch, err := productsmodel.GetBestMatch(ctx, ctx.transaction, getSourceString(namespace, ctx.aggregatedBucket.Organization), ctx.billingDate)
+// getBestMatch tries to get the best match for product, discount and query
+func (d *Database) getBestMatch(dctx *Context) error {
+	namespace := *dctx.namespace
+	productMatch, err := productsmodel.GetBestMatch(dctx, dctx.transaction, d.sourceString.getSourceString(), d.BillingDate)
 	if err != nil {
 		return fmt.Errorf("cannot get product best match for namespace %s: %w", namespace, err)
 	}
-	ctx.product = productMatch
+	dctx.product = productMatch
 
-	discountMatch, err := discountsmodel.GetBestMatch(ctx, ctx.transaction, getSourceString(namespace, ctx.aggregatedBucket.Organization), ctx.billingDate)
+	discountMatch, err := discountsmodel.GetBestMatch(dctx, dctx.transaction, d.sourceString.getSourceString(), d.BillingDate)
 	if err != nil {
 		return fmt.Errorf("cannot get discount best match for namespace %s: %w", namespace, err)
 	}
-	ctx.discount = discountMatch
+	dctx.discount = discountMatch
 
-	queryMatch, err := queriesmodel.GetByName(ctx, ctx.transaction, queryAndZone)
+	queryMatch, err := queriesmodel.GetByName(dctx, dctx.transaction, d.sourceString.getQuery())
 	if err != nil {
 		return fmt.Errorf("cannot get query by name for namespace %s: %w", namespace, err)
 	}
-	ctx.query = queryMatch
+	dctx.query = queryMatch
 
 	return nil
 }
 
-func (d *Database) adjustStorageSizeUnit(ctx *transactionContext) error {
-	var quantity float64
-	if query.Unit == defaultUnit {
-		quantity = ctx.aggregatedBucket.StorageUsed / 1024 / 1024 / 1024
-	} else {
-		return fmt.Errorf("unknown query unit %s", query.Unit)
+func (d *Database) saveFacts(dctx *Context) error {
```

Comment:
Due to the aggregation we'll probably lose some data here. We said, that we bill a database for the whole hour if it's in any of the metrics runs.

If we don't aggregate it's very simple, we just write a 1 into the database, if we see an instance at any given hour, easy.

But we aggregate thing here, so if we run, let's say every 10 minutes, it could happen that the customer has 3 running instances for the given hour. 10 minutes later he deleted one instance and the metrics only return 2. We lose the billing for one instance.

So on each run we have to determine for the given hour, if the value that was written before is higher or lower and only write the higher value.

TL;DR: check if there's already a value for the current hour, if the existing value is smaller, overwrite it, if it isn't keep it.

Reply:
I have updated the code and introduced the function `isFactsUpdatable` which will make sure that fact table is updated only for DBaaS service. Also the record will only be saved if the quantity is higher than the previous value. It will never save a lower value. Moreover, we have to update the [library](https://github.com/vshn/cloudscale-metrics-collector/tree/master/pkg) of the billing DB to support updating facts table. It is already [implemented](https://github.com/vshn/cloudscale-metrics-collector/blob/master/pkg/productsmodel/products.go#L75) for products.

Reply:
Created a [PR](https://github.com/vshn/cloudscale-metrics-collector/pull/29) in the library code.